libtransmission/peer-mgr.h

```diff
@@ -46,6 +52,329 @@ enum
+    {
+        connection_changed_at_ = now;
+
+        is_connected_ = is_connected;
+
+        if (is_connected_)
+        {
+            num_consecutive_fails_ = {};
+            piece_data_at_ = {};
+        }
+    }
+
+    [[nodiscard]] constexpr auto is_connected() const noexcept
+    {
+        return is_connected_;
+    }
+
+    // ---
+
+    [[nodiscard]] bool is_blocklisted(tr_session const* session) const;
+
+    void set_blocklisted_dirty()
+    {
+        blocklisted_.reset();
+    }
+
+    // ---
+
+    constexpr void ban() noexcept
+    {
+        is_banned_ = true;
+    }
+
+    [[nodiscard]] constexpr auto is_banned() const noexcept
+    {
+        return is_banned_;
+    }
+
+    // ---
+
+    [[nodiscard]] constexpr auto connection_attempt_time() const noexcept
+    {
+        return connection_attempted_at_;
+    }
+
+    constexpr void set_connection_attempt_time(time_t value) noexcept
+    {
+        connection_attempted_at_ = value;
+    }
+
+    constexpr void set_latest_piece_data_time(time_t value) noexcept
+    {
+        piece_data_at_ = value;
+    }
+
+    [[nodiscard]] constexpr bool has_transferred_piece_data() const noexcept
+    {
+        return piece_data_at_ != time_t{};
+    }
+
+    [[nodiscard]] constexpr auto reconnect_interval_has_passed(time_t const now) const noexcept
+    {
+        auto const interval = now - std::max(connection_attempted_at_, connection_changed_at_);
+        return interval >= get_reconnect_interval_secs(now);
+    }
+
+    [[nodiscard]] constexpr std::optional<time_t> idle_secs(time_t now) const noexcept
+    {
+        if (!is_connected_)
+        {
+            return {};
+        }
+
+        return now - std::max(piece_data_at_, connection_changed_at_);
+    }
+
+    // ---
+
+    constexpr void on_connection_failed() noexcept
+    {
+        if (num_consecutive_fails_ != std::numeric_limits<decltype(num_consecutive_fails_)>::max())
+        {
+            ++num_consecutive_fails_;
+        }
+    }
+
+    [[nodiscard]] constexpr auto connection_failure_count() const noexcept
+    {
+        return num_consecutive_fails_;
+    }
+
+    // ---
+
+    constexpr void set_pex_flags(uint8_t pex_flags) noexcept
+    {
+        if ((pex_flags & ADDED_F_CONNECTABLE) != 0U)
+        {
+            set_connectable();
+        }
+
+        if ((pex_flags & ADDED_F_UTP_FLAGS) != 0U)
+        {
+            set_utp_supported();
+        }
+
+        is_seed_ = (pex_flags & ADDED_F_SEED_FLAG) != 0U;
+    }
+
+    [[nodiscard]] constexpr uint8_t pex_flags() const noexcept
+    {
+        auto ret = uint8_t{};
+
+        if (is_connectable_ && *is_connectable_)
+        {
+            ret |= ADDED_F_CONNECTABLE;
+        }
+
+        if (is_utp_supported_ && *is_utp_supported_)
+        {
+            ret |= ADDED_F_UTP_FLAGS;
+        }
+
+        if (is_seed_)
+        {
+            ret |= ADDED_F_SEED_FLAG;
+        }
+
+        return ret;
+    }
```

Comment:
We don't use `ADDED_F_ENCRYPTION_FLAG` in our logic right now, but I think it should still be stored in `tr_peer_info` for the sake of propagating the info through PEX.

Reply:
Same for `ADDED_F_HOLEPUNCH`. It is a regression from the previous behaviour.

...Now that I wrote this comment, I reckon we should store the PEX flag as a bitfield just like before inside `tr_peer_info`, so that all flags unknown to us will be passed over to other peers transparently.

Reply:
Good point.

Reply:
Added back in 3eefcd15921781deb5fff2c225cd9bb646cafdd3

Reply:
Just need to add `ADDED_F_HOLEPUNCH` to `tr_peer_info` as well and then this PR will LGTM. 👍🏻

Reply:
ahh you beat me to it. I was thinking the same thing and just pushed it up in efd6e2cd4ac5c76b7fc0be7c13af784c714560b2